erts/emulator/beam/erl_bif_lists.c

```diff
@@ -146,93 +157,538 @@ BIF_RETTYPE append_2(BIF_ALIST_2)
+        erts_free(ERTS_ALC_T_TMP, (void *) context->copy_of_b);
+        context->copy_of_b = NULL;
+    }
+    return 1;
+}
+
+
+/* Called when magic binary with ErtsSubtractContext is destroyed */
+static int
+subtractcontext_binary_dtor(Binary *context_bin) {
+    ErtsSubtractContext *ctx = ERTS_MAGIC_BIN_DATA(context_bin);
+    return subtractcontext_dtor(ctx);
 }
 
+
+/* Prepare magic ref to magic binary with our state.
+ * The magic binary with context in it should already be created.
+ * Return a triple {A, B, MagicRef}
+ */
+static ERTS_INLINE Eterm
+subtract_create_returnstate(Process *p, Eterm A, Eterm B, Binary *context_b) {
+    const static int TUPLE3_SIZE = 3 + 1;
+    Eterm *hp = HAlloc(p, ERTS_MAGIC_REF_THING_SIZE + TUPLE3_SIZE);
+    Eterm state_binref = erts_mk_magic_ref(&hp, &MSO(p), context_b);
+    return TUPLE3(hp, A, B, state_binref);
+}
+
+
+/* Iterate over A and for each element that is present in B, skip it
+ * in A and remove it from B (by moving last element of B in its place and
+ * making B shorter), otherwise that element is copied to the result.
+ * Repeat until either end of A is reached or B is empty.
+ *
+ * Result is built in s->result_array of Eterm[result_size]
+ * Returns: 0 if the work is not done and must call again
+ * Returns: 1 if the work is done and must copy result to a new list
+ */
+static int
+subtract_trappable_scan(Process *p, Eterm A, ErtsSubtractContext *context) {
+    ErtsSubtractScanContext *s = &context->s.scan;
+    Eterm iter_a = s->iter_a; /* load iterator from state */
+    static const Uint WORK_BEFORE_TRAPPING = IF_DEBUG(30, 300);
+    register Uint reds = WORK_BEFORE_TRAPPING;
+    Eterm *b = context->copy_of_b;
+
+    /* A is guaranteed to be proper list, otherwise erts_list_length() call at
+     * start will fail */
+    while (is_list(iter_a)) {
+        const Eterm *ptr_a  = list_val(iter_a);
+        Sint found_at = subtract_find_in_array(CAR(ptr_a), b, context->len_b);
+        if (found_at >= 0) {
+            /* Remove elem from B and skip; shrink B by 1 */
+            context->len_b--;
+            b[found_at] = b[context->len_b];
+
+            if (context->len_b == 0) {
+                iter_a = CDR(ptr_a);
+
+                /* B is empty at this point, so copy remaining A elements */
+                while (is_list(iter_a)) {
+                    ptr_a = list_val(iter_a);
+                    context->result_array[s->result_size] = CAR(ptr_a);
+                    s->result_size++;
+                    iter_a = CDR(ptr_a);
+                }
+                break; /* nothing left in B, end search here */
+            }
+        } else {
+            context->result_array[s->result_size] = CAR(ptr_a);
+            s->result_size++;
+        }
+        iter_a = CDR(ptr_a);
+        reds--;
+    }
+
+    if (reds) {
+        /* Reds not 0, work seems to be completed */
+        return am_true;
+    }
+    s->iter_a = iter_a; /* save iterator */
+    return am_false;
```

Comment:
Doesn't this function return `int`?

Reply:
Eterm is an int and compiler was ok with this heresy. Thanks, fixed.

Reply:
Changing this to 1 and respectively 0 suddenly make algorithm 7 times faster on random inputs of 500 elements each. i think i broke something here.

Reply:
Yes, a bug in result building. Brb to eat something and fixing